### DISCUSSION TITLE
DAGs vanishing from serialized_dag table

### DISCUSSION THREAD

**Discussion by kgweber-sag** [apache/airflow#22631] [Q&A] [answered]
I am using Airflow 2.2.4 in the Docker configuration descriped in the [documentation](https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html). 

When I add my own DAGs to the `dags` directory, Airflow picks them up and starts working on them just fine, but the UI intermittently puts up a red error, `DAG "qualtrics_extract" seems to be missing.`, and if I happen to ask to start a job or view a graph, sometimes it goes all the way to an exception complaining about one of a variety of `NoneType` errors.

The Airflow scheduler will throw errors like this: 

`[2022-03-30 18:25:53,714] {dagbag.py:205} WARNING - Serialized DAG qualtrics_extract no longer exists` and then go quiet, until the DAG runs. Then it starts streaming, over and over and over,

`airflow-scheduler_1  | [2022-03-30 18:30:28,966] {scheduler_job.py:1016} ERROR - Couldn't find dag qualtrics_extract in DagBag/DB!`

`airflow-scheduler_1  | [2022-03-30 18:30:28,970] {scheduler_job.py:824} ERROR - DAG 'qualtrics_extract' not found in serialized_dag table` 

The DAG keeps running, but the UI is useless.

When I look under the hood, DAGs are coming and going from the `serialized_dag` table at an alarming rate. They're randomly vanishing and returning, and a missing DAG there seems to be correlated with UI problems

I've blown this configuration away and rebuilt it from the files provided in the documentation several times, and I've taken this configuration to a completely different machine with a completely different OSes. (MacOS and RHEL 7 are both doing this). I'm opening a Discussion here because I don't have a reproducible example I can define - this is really out of the box except I have my DAGs in it.

Can those of you who understand the scheduler help me troubleshoot further? It feels as though there are two jobs at war with each other somehow, creating and destroying `serialized_dag` entries, but I could be mistaken. How do I figure out what's going on?

**Comment by potiuk** [comment, marked as answer]
First of all I really appreciate that you opened a discussion as advised :). This makes it easy for us to distinguish real issues from some forms of troubleshooting.

t looks like your DAGs when parsed might produce different or dynamic results. You shoud make sure that your DAGs always produce pretty similarly looking  set of DAGs every single time it is parsed. Look at any cases where whatever is produced by parsing a DAG depends on some external factors - database availablity, pulling some configuration etc. In fact it is best to make suret that whatever you need is locally available in the DAGs folder (for example json describing your DAG configuration). Also make sure that the filesystem your DAG is on is "stable"  - i.e. that you will always get a consistent set of files there when you scan it. If you update DAG files elsewhere and try to distribute them with the filesystem- make sure whatever changes come there are atomic. I.e. no half-transmitted files, no in-progress writing. Also maybe simply you have some intermittent flakiness during parsing - some errors that depend on racing condtions. Those are the worst to track - but usually you can avoid it by getting everything done "locally" (i.e don't reach out to remote sources when you parse your DAG).

Finally make sure that you do not have a second scheduler runnig in paralllel accessing your database and having different source for DAGs. THIS SCENARIO HAPPENED IN THE PAST. - so however unlikely it is, if you have a database that is remotely accessible, check if by any chance there is no test or dev server by mistake confgured to use your prod db (happened, really :) ).

You can see more concrete advices and examples here: https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code  - this and next chapter have a lot of advices you can follow.

**Reply by kgweber-sag** [reply]
Thank you so much for your quick reply! This is part of how I've become stuck, because I've seen a lot of chatter about dynamic DAGs causing trouble, and I don't think we're doign that at all. We're very simple-minded about this stuff so far. This is typical of what we build:

```
from airflow import DAG
from airflow.providers.http.operators.http import SimpleHttpOperator
from customoperators.jsontomysqloperator import JsonToMySQLOperator
from datetime import datetime
from models.assetmodels import OttoAsset, AssetBase

default_args = {
    "owner": "airflow",
    "email": ["halp-halp-halp@our-u.edu"],
    "email_on_failure": True,
    "email_on_retry": False,
    "retries": 1,
}

dag = DAG(
    "collect_otto_data",
    description="Brings current Otto data into a dbms table",
    schedule_interval="30 5 * * *",
    start_date=datetime(2022, 2, 1),
    catchup=False
)

get_otto_json_task = SimpleHttpOperator(
    task_id="get_otto_json_task",
    http_conn_id="otto_api",
    endpoint="systems.json",
    data={"filter_organization_id": "**"},
    method="GET",
    headers={"Content-Type": "application/json"},
    dag=dag,
)

db_write_task = JsonToMySQLOperator(
    task_id="db_write_task",
    model=OttoAsset,
    base=AssetBase,
    mysql_conn_id="[that-server]",
    action="replace",
    pushing_task="get_otto_json_task",
    dag=dag,
)

get_otto_json_task >> db_write_task
```
All this resides in a directory local to the `docker-compose.yaml` file, which is attached to `/dags` via a `volumes:` call

The database is just the little postgres database that the sample docker-compose file stands up. Nothing else even *can* talk to it.

Am I doing something fundamentally, terribly wrong?

**Reply by potiuk** [reply]
Have you seen if your containers are continuously restarting ? That also might explain things (Maybe you do not have enough memory)

Note that Docker Compose is really a developer-tool and I personally hate it by the way how it "pretends" to be easy to run containers, while you REALLLY have to go into details and understand what's going on if you want to do anything serious. So if you "expect" from the docker compose to provide you the layer of abstraction that frees you from understanding in detail what's going on in detail - you are wrong. K8S + our helm chart does much better job in that (though it is more complex) as advised.

You NEED to learn how to diagnose it. You NEED to know about networking. You NEED to understand how your volumes are allocated on your system. You need to understand Docker, Containers and how they interact if you want to go Docker Compose route. We put warnings in our docs.

(This is a bit rant and a bit warning).

**Reply by kgweber-sag** [reply]
Understood and appreciated. I'm just a developer type with a reasonable, if not deep, handle on Docker. If I stand up a K8S cluster right now, my sysadmins will hurt me. Trust me, I was tempted.

The containers are happy and purring along, which is why I'm chasing the idea that the scheduler is somehow at war with itself inside its container.

Is there anything boneheaded I'm doing in that DAG? It's calling modules that are resident in subdirectories of the /dags folder, but that worked well for us on version 1.10.

**Reply by josh-fell** [reply]
FWIW, related to the top-level code, this _may_ also be a symptom of what's happening in the constructor of `JsonToMySQLOperator`. If that custom module is connecting to a database or some heavy computation, etc. could make the parsing unstable.

**Reply by potiuk** [reply]
Yep. We have no insigth on the "custom operator" and what it is doing. This falls into "doing stuff in Top-Level Code"

**Reply by kgweber-sag** [reply]
It is connecting to a database - many of our customoperators do.

This is a helpful track to follow and I'll try restructuring this chap to see if I get somewhere.

I'll go back to the best practices dox that @potiuk suggests. Generally speaking, should this sort of code be built into a plug-in instead?

```
from airflow.models.baseoperator import BaseOperator
from airflow.providers.mysql.hooks.mysql import MySqlHook
from sqlalchemy.orm import sessionmaker, class_mapper, ColumnProperty

import json


class JsonToMySQLOperator(BaseOperator):
    """
    A reasonably flexible operator that will bring in JSON that has been converted into a list of dicts
    with json_loads, and whose date formats have been normalized, and then load it into a MySQL table.
    The code will build the table according to the sqlalchemy model you specify, and will filter
    the json to only include those columns in the model.

    It's on you, should you choose to append data, to make sure your primary keys are either being
    autoincremented in the model, or else guaranteed unique before you insert.
    """

    def __init__(
        self,
        model,  # this is a sqlalchemy model
        base,  # this is a sqlalchemy base
        action: str = "append",  # or 'replace'
        mysql_conn_id: str = None,
        pushing_task=None,
        *args,
        **kwargs,
    ) -> None:
        super().__init__(*args, **kwargs)
        self.model = model
        self.base = base
        self.action = action
        self.mysql_conn_id = mysql_conn_id
        self.pushing_task = pushing_task

    def execute(self, context):
        m_hook = MySqlHook(mysql_conn_id=self.mysql_conn_id)
        sql_alchemy_engine = m_hook.get_sqlalchemy_engine()  # engine_kwargs=engine_kwargs)

        Session = sessionmaker()
        Session.configure(bind=sql_alchemy_engine)
        mysql_session = Session()
        task_instance = context["task_instance"]
        object_list = json.loads(task_instance.xcom_pull(task_ids=self.pushing_task))
        object_list = filter_objects(self.model, object_list)
        if self.action == "replace":
            mysql_session.execute(f"DROP TABLE IF EXISTS {self.model.__tablename__}")
        self.base.metadata.create_all(sql_alchemy_engine)
        to_import = [self.model(**x) for x in object_list]
        mysql_session.add_all(to_import)
        mysql_session.commit()
        return


def filter_objects(model, object_list):
    object_fields = [prop.key for prop in class_mapper(model).iterate_properties if isinstance(prop, ColumnProperty)]
    objects_to_return = list()
    for obj in object_list:
        objects_to_return.append({key: value for key, value in obj.items() if key in object_fields})
    return objects_to_return
```

**Comment by kgweber-sag** [comment]
I want to follow up by letting you all know that your advice was very helpful.

What I did:

I burned everything down and rebuilt some of my operators with the expensive imports moved inside working code blocks, and (more important) I built out modules in the `plugins` directory and imported from there. Everything is working just beautifully now and I'm grateful for all your advice and pointers in the right direction.

**Reply by potiuk** [reply]
Glad it worked ! Thanks for information :)
